vispy/visuals/_scalable_textures.py

```diff
@@ -295,29 +295,25 @@ def clim_normalized(self):
 
     @staticmethod
     def _scale_data_on_cpu(data, clim, copy=True):
-        if copy:
-            should_cast_to_f32(data.dtype)
-            data = np.array(data, dtype=np.float32, copy=copy)
-        elif not copy and not np.issubdtype(data.dtype, np.floating):
-            raise ValueError("Data must be of floating type for no copying to occur.")
-
+        data = np.array(data, dtype=np.float32, copy=copy)
         if clim[0] != clim[1]:
             data -= clim[0]
-            data *= 1.0 / (clim[1] - clim[0])
-        if should_cast_to_f32(data.dtype):
-            data = data.astype(np.float32)
+            data *= 1 / (clim[1] - clim[0])
         return data
 
     def scale_and_set_data(self, data, offset=None, copy=True):
         """Upload new data to the GPU, scaling if necessary."""
+        # ensure dtype is the same as it was before, or funny things happen
+        # we use the `copy` value only here, to avoid copyign multiple times when True
+        data = np.array(data, dtype=self._data_dtype, copy=copy)
```

Comment:
But this was true before as well; actually it was worse, because if `copy` was `True`, the array was being copied *multiple times*. See how `copy=copy` was passed twice in the code below, and was being propagated again to `self.set_data(..., copy)`, which in turn uses it again. I think currently the data is copied 2, maybe 3 times if `copy=True`. With this change, it will only happen once (here) if `True`.

Reply:
Yeah and I guess if copy is False then it won't actually do the copy so if a copy is required by the logic below it will just happen. Got it.